### DISCUSSION TITLE
authorizeResource alternatives

### DISCUSSION THREAD

**Discussion by uikolas** [laravel/framework#50552] [Q&A]
After upgrading to Laravel 11 I noticed that 11 version dropped documentation for resource authorization [link to 10 version resource docs](https://laravel.com/docs/10.x/authorization#authorizing-resource-controllers) and [in 11 version it does not exists](https://laravel.com/docs/11.x/authorization) 

And as new version propose to use [clean base controller](https://laravel.com/docs/11.x/releases#base-controller-class), but we can still use `AuthorizesRequests` trait. 
But then [`authorizeResource`](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Foundation/Auth/Access/AuthorizesRequests.php#L83) method stops working, since it does have access to [`middleware` function](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Routing/Controller.php#L23) anymore

So maybe a new handy approach exist to set Policy on resource controller?

I know I can extend the base controller. But since the docs are removed I guess a new approach should exist?
Also I know I can add on each resource method, to check if user is authorized. But maybe we can do it more simpler?

**Comment by Khazl** [comment]
I have noticed the same thing and am facing the same problem.
Is there an official best practice for this?

**Reply by MorganGonzales** [reply]
I assume that Laravel 11 is heading towards being specific in defining authorization rules, so probably having a `Gate::authorize` definition for each controller action, instead of using the `authorizeResource` which is very global. And I think this is the best practice for Laravel 11.

**Reply by Khazl** [reply]
> having a Gate::authorize definition for each controller action

Doesn't necessarily sound like a good alternative and a planned best practice. 
That's exactly why the question comes up. Can anyone officially confirm that this is the case?

**Comment by i6media** [comment]
You can still extend the `\Illuminate\Routing\Controller` instead of `App\Http\Controllers\Controller` and use the `\Illuminate\Foundation\Auth\Access\AuthorizesRequests\AuthorizesRequests` in your custom controllers in Laravel 11:

```
<?php

class FooController extends \Illuminate\Routing\Controller
{
    use \Illuminate\Foundation\Auth\Access\AuthorizesRequests;

    public function __construct()
    {
        $this->authorizeResource(FooResource::class, 'parameter');
    }
}
```

**Reply by hesham-fouda** [reply]
In this case you can't use `Illuminate\Routing\Controllers\HasMiddleware`

**Reply by jalexmelendez** [reply]
Thank you!!! I use this feature heavily and I am updating an application from laravel 10 to 11 and ran to this issue

**Comment by GautierDele** [comment]
@taylorotwell is there a particular reason you removed this in the documentation on this commit ?
https://github.com/laravel/docs/commit/745bab30b0c845db5bacadcc99309a8d1d6565c3

Just to be sure this is not about to be deprecated 😄

**Reply by heitorvrb** [reply]
This is the question we want to be clarified. @taylorotwell

**Comment by aplauche** [comment]
This is really disappointing if it is being removed. Needing to specify for each method in the controller negates a big benefit of resource controllers. Without authorizeResource it feels better to manually specify each route and use middleware. This feels like a critical component of the idea of a "resource".

**Reply by crynobone** [reply]
There no where in the upgrade guide saying `Illuminate\Http\Controller` was deprecated, it just not being configured as the default.

**Reply by aplauche** [reply]
Got it, just saw the comment above that mentioned deprecation, and also wanted to clarify whether that was the case, but happy to hear it is just a defaults change! Will continue to just use http controller. Thanks!

**Comment by osbre** [comment]
Honestly, I don't see why it makes a big difference. Just copy and paste `Gate::authorize` a few more times or make it part of your customized generator stub.

Not having `authorizeResource` makes it more straightforward to figure out what's happening. Previously we also had to think whether there is `authorizeResource` being applied, now we can simply look at the method itself.

Anyone else agree? 😄

Upd: for new projects, I mean.

**Reply by francoism90** [reply]
@osbre After taking a look at the previous Laravel 9.x [Controller class](https://github.com/laravel/laravel/blob/9.x/app/Http/Controllers/Controller.php), I can now agree with you. :)

The problem is when having something like Livewire. The 9.x Controller class, has a few traits that provide many methods, that can also be used directly instead - because of the nature of Livewire. This makes the controller class expose public methods and variables, which can become a security issue.

**Comment by f-liva** [comment]
I don't understand why Laravel has become more complex than it used to be. All these features were in line with the concept of "Web framework for Artisans," but now instead of simplifying, Laravel is starting to complicate things.

**Comment by mayanksdudakiya** [comment]
The method still exist as it is. There is no major change in the file: https://github.com/laravel/framework/blob/12.x/src/Illuminate/Foundation/Auth/Access/AuthorizesRequests.php#L87-L106

**Reply by francoism90** [reply]
This seems to be related to the removable of the default `Controller` class: https://github.com/laravel/laravel/blob/9.x/app/Http/Controllers/Controller.php
